EgressIP: miscellaneous fixes

alexanderConstantinescu opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Trivial changes are exempt from following this template.
If your change is non-trivial, please provide the following information:
-->

This fixes some problems found for the implementation and refactor made in https://github.com/ovn-org/ovn-kubernetes/pull/2691

**- What this PR does and why is it needed**
<!--
A summary of the changes within this pull request and some context
as to why they were made
-->

**- Special notes for reviewers**
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->


**- How to verify it**
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

alexanderConstantinescu commented:
/retest

kyrtapz commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -123,14 +128,24 @@ func (oc *Controller) reconcileEgressIP(old, new *egressipv1.EgressIP) (err erro
+			exists := false
+			for _, status := range statusToRemove {
+				if reflect.DeepEqual(status, oldStatus) {
+					exists = true
```
NIT: can we just move `statusToRemove = append(statusToRemove, oldStatus)` here and not use `exists`?

alexanderConstantinescu replied:
Not really because of what I mentioned in the doc bloc above:
```
        // ....If a user removes
	// an egress IP from .spec.egressIPs we however need to guard against adding
	// the same status to statusToRemove twice, which would happen with the code
	// above for validStatuses because both newEIP and oldEIP will have the old
	// assignment status.
```

If an update comes in: 

old: 

```yaml
apiVersion: k8s.ovn.org/v1
kind: EgressIP
metadata:
    name: egressip
spec:
    egressIPs:
    - 10.0.208.210
    - 10.0.208.211
    namespaceSelector:
        matchLabels:
            name: test
status:
    items:
    - EgressIP: 10.0.208.210
      Node: node1
    - EgressIP: 10.0.208.211
      Node: node2
```

new: 

```yaml
apiVersion: k8s.ovn.org/v1
kind: EgressIP
metadata:
    name: egressip
spec:
    egressIPs:
    - 10.0.208.210
    namespaceSelector:
        matchLabels:
            name: test
status:
    items:
    - EgressIP: 10.0.208.210
      Node: node1
    - EgressIP: 10.0.208.211
      Node: node2
```

In which case, without this check: `statusToRemove` would contain duplicate items for `    - EgressIP: 10.0.208.211
      Node: node2`

alexanderConstantinescu replied:
sets can only be comprised of comparable primitive types, and I was going to say that this applies to map keys as well..but I learned something new today: a map can be comprised of a struct if all the struct fields are comparable, this is the case for an `EgressIPStatusItem`, so I will change this  to a  map based implementation.

kyrtapz reviewed (approved):
/lgtm

flavio-fernandes reviewed (commented):
/lgtm

trozet commented on go-controller/pkg/ovn/pods.go:
```diff
@@ -443,47 +443,24 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) (err error) {
+		// update performed below. Doing this will ensure that we won't race by
+		// having the SNAT created here, after the performed removal by the
+		// egress IP code.
+		if err := oc.addPodGatewayRoutes(pod, podIfAddrs); err != nil {
```
I'm not sure this is the right solution. Moving this code here under if needsIP means that the routes are only added if this is a new IP assignment, where as before it would happen regardless. I think trying to modify pods.go function is a recipe for more bugs and we shoudl use proper locking/signaling/handling between egress IP handler and pod handler to ensure no race.

alexanderConstantinescu replied:
No, see L463. We add this if `!needsIP`

alexanderConstantinescu replied:
And again, this doesn't have anything to do with locking. The "racy-ness" of concern here is strictly in terms of the logical order of which things are executed. The egress IP code needs to be sure that the pod GR SNAT is created before the annotation is applied, before that wasn't the case.

alexanderConstantinescu replied:
Actually, why do we always add the pod GW SNAT? I realize now that this is not going to work with egress IP. They will always compete for every pod update, and with the change in this PR the egress IP code will not work because we'll continuously add the pod GW SNAT for every pod update, whereas the egress IP code only removes it once the pod has its IPs assigned.

alexanderConstantinescu replied:
The one case I figured was covered by always adding the pod GW SNAT was in case someone deletes the NB DB and restarts ovnkube-master. The assumption then is that we should have the gateway routes re-created, even though the pod already has the annotation. I hence added this to `syncPods`. We are thus creating the pod gateway routes whenever a pod does not have an IP assignment yet, or when we restart, always.

trozet replied:
Re-iterating offline discussion: Alex to look into if we could use the port cache as our signal to know addLogicalPort is finished and then egressIP knows it can remove/add the new SNAT.

alexanderConstantinescu replied:
Port cache works just fine and resolves our concurrency issue with addLogicalPort. I've updated my patch

trozet commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -398,9 +398,15 @@ func (oc *Controller) reconcileEgressIPPod(old, new *v1.Pod) (err error) {
-	// add/remove the setup accordingly. Pods cannot match multiple EgressIP
-	// objects: that is considered a user error and is undefined. Hence, just
-	// just perform the setup against the first object matched.
+	// add/remove the setup accordingly. Pods should not match multiple EgressIP
```
this commit needs a commit message with a good explanation, probably something summarizing this comment

alexanderConstantinescu replied:
OK, will add that

trozet commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1641,6 +1628,9 @@ func (e *egressIPController) getGatewayRouterJoinIP(node string, wantsIPv6 bool)
 }
 
 func (e *egressIPController) getPodIPs(pod *kapi.Pod) ([]*net.IPNet, error) {
+	if pod.Spec.HostNetwork {
```
seems like this is not part of this commit

trozet commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -513,23 +514,33 @@ func (oc *Controller) reconcileCloudPrivateIPConfig(old, new *ocpcloudnetworkapi
+		// a CloudPrivateIPConfig delete because the EgressIP has been deleted
+		// then we need to remove the setup made for it, but not update the
+		// object.
+		egressIP, err := oc.kube.GetEgressIP(egressIPName)
```
use informer instead of kubeclient

alexanderConstantinescu replied:
See my above comments on the informer cache and multiple updates in a row for objects which we read and write from/to. I would rather use a extra network call than spend days debugging a difficult cache problem.

trozet commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -164,11 +164,15 @@ func (oc *Controller) reconcileEgressIP(old, new *egressipv1.EgressIP) (err erro
+		// on restart. We won't perform any additional transactions to the NB DB
+		// for things which exists because the libovsdb operations use
+		// modelClient which is idempotent.
+		if err := oc.addEgressIPAssignments(name, statusToKeep, newEIP.Spec.NamespaceSelector, newEIP.Spec.PodSelector); err != nil {
```
should this commit be a fixup/squashed to your previous commit where you fix cache sync on start up?

alexanderConstantinescu replied:
Yeah, will fix that.

alexanderConstantinescu commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -988,29 +988,24 @@ func (oc *Controller) isAnyClusterNodeIP(ip net.IP) *egressNode {
 	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
 		// Get the latest item from the store before updating.
-		egressIP, err := oc.watchFactory.GetEgressIP(name)
+		egressIP, err := oc.kube.GetEgressIP(name)
```
This doesn't really have anything to do with guarding against races in our own code. Keep in mind that this is already done by means of the `egressIPAssignmentMutex`. The reason for this change is to make sure we don't act on stale data when we need to perform multiple updates in a row. Consider the following:

You have an egress IP object with status: [1, 2], you want to add 3 and 4 sequentially using this function. If we are using the informer cache, so ending up in the following situation takes little to no effort:

1. `updateEgressIPStatusWithItem` is called, adding 3 to [1,2]. The informer cache is up-to-date so the object retrieved does have status [1,2]
2. `updateEgressIPStatusWithItem` is called again, adding 4 to [1,2,3]. The informer cache is not up-to-date because it still hasn't been updated by the update made in 1., so the object returned has status [1,2]
3. The update in 1. is overwritten by the one made in 2. leading to a final state of: [1,2,4]

You can't rely on the informer cache when you are both reading and writing to the same object and field. You can when only reading and you have your own separate data store to update/write to.

trozet replied:
Just re-iterating our offline discussion here. Alex is going to look into using Patch operation instead of Update to see if that will be good enough for adding/removing EgressIPStatusItems.

alexanderConstantinescu replied:
Using a patch operation saves us from retrieving the object just before we update it, which we did originally because we needed to be sure that we didn't overwrite any user changes for updates which might take a long time. Given that patch operations can specify which individual field to update, we don't have this problem anymore and do not need to retrieve the object just before updating it. 

As to removing all client calls used by the egress IP code, that can't be solved by this. Patch operations are limited by  [RFE 6902](https://datatracker.ietf.org/doc/html/rfc6902#appendix-A.4) which only allows removal of slice elements by the index, hence we can't really use it since that model isn't "concurrency-safe". An example of this is: say we have a status [1,2,3] and we want to remove [2,3]. To do that we'd need to remove index 1 first and then 2, but by the time 1 has been removed, 2 is not at index 2 anymore. 

We still need the `WatchEgressNodes` and `reconcileCloudPrivateIPConfig` functions to retrieve the last state of the egress IP object and pass that to the `reconcileEgressIP` so that we  are sure to always update the latest version.

alexanderConstantinescu commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -784,28 +787,27 @@ func (oc *Controller) deleteAllocatorEgressIPAssignments(statusAssignments []egr
+// (egress IP name - status) basis. The idea is thus to list the full content of
+// the NB DB for that egress IP object and delete everything which match the
+// status. We also need to update the podAssignment cache and finally re-add the
+// ICNI setup in case the pod still exists.
```
I intertwined the commits. This part of the doc bloc should have gone with a later commit...

alexanderConstantinescu commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -169,6 +169,10 @@ func (oc *Controller) reconcileEgressIP(old, new *egressipv1.EgressIP) (err erro
+		// Add all assignments which are to be kept to the allocator cache,
+		// allowing us to track all assignments which have been performed and
+		// avoid incorrect future assignments due to a de-synchronized cache.
+		oc.addAllocatorEgressIPAssignments(name, statusToKeep)
```
No, not really: `addEgressIPAssignments` was only called if `len(ipsToAdd) > 0`, which would not always be the case if we restart since `ipsToAdd` contain the diff between what needs to be assigned and that which is already assigned. We need to everything existing to the allocator cache though as to keep the state up-to-date. I will add a better commit message though

alexanderConstantinescu commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -535,14 +543,17 @@ func (oc *Controller) reconcileCloudPrivateIPConfig(old, new *ocpcloudnetworkapi
 	egressIPString := cloudPrivateIPConfigNameToIPString(cloudPrivateIPConfigName)
-	eIPs, err := oc.watchFactory.GetEgressIPs()
+	// retrieve the EgressIPs from the API server. We have multiple go-routines
+	// modifying the state of the egress IP object, hence: we need to ensure
```
As mentioned on that other comment, nothing to do with locking in our own code.

alexanderConstantinescu replied:
In any case this function is removed completely as a part of https://github.com/ovn-org/ovn-kubernetes/pull/2734/commits/bf770975ac6e160e0b091c07547de14f58fa258e 

I could maybe squash both commits ?

alexanderConstantinescu replied:
I squashed both commits.

alexanderConstantinescu commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -535,14 +543,17 @@ func (oc *Controller) reconcileCloudPrivateIPConfig(old, new *ocpcloudnetworkapi
 // egress IP, hence why this should work.
 func (oc *Controller) getEgressIPFromCloudPrivateIPConfigName(cloudPrivateIPConfigName string) (*egressipv1.EgressIP, error) {
 	egressIPString := cloudPrivateIPConfigNameToIPString(cloudPrivateIPConfigName)
-	eIPs, err := oc.watchFactory.GetEgressIPs()
```
Logically it does fit in both commits since this does apply to the CloudPrivateIPConfig part. But I can move it to the other one.

alexanderConstantinescu commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -791,6 +791,9 @@ func (oc *Controller) addPodEgressIPAssignments(name string, statusAssignments [
 	existing, exists := oc.eIPC.podAssignment[podKey]
 	if !exists {
 		remainingAssignments = statusAssignments
+		if err := oc.eIPC.deletePerPodGRSNAT(pod); err != nil && !errors.Is(err, skippedPodError) {
```
> Why would we need to delete this here when if !exists?

`oc.eIPC.podAssignment` tracks what egress IP status has been setup for each pod which matches an egress IP. If nothing exists for a pod, it means nothing has been performed yet and we are going to now. Hence, we need to delete the pod GR SNAT and set up the egress IP SNAT. 

> Why not just always remove an invalid pod GR SNAT or update it during egress IP add?

Because this function will end up getting called from `reconcileEgressIP`, `reconcileEgressIPNamespace`, `reconcileEgressIPPod` for each pod and for each pod update. On restart we might end up here in the order of tens for every single pod, which, even though things should be idempotent, is a bit excessive on the CPU when there is no need to do so.

alexanderConstantinescu commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1553,23 +1549,25 @@ func (e *egressIPController) deletePodEgressIPAssignment(egressIPName string, st
-				if err != nil {
-					return fmt.Errorf("unable to retrieve pod IPs, err: %v", err)
-				}
+			podIPs, err := e.getPodIPs(pod)
```
The util function does not have the same implementation as this one, this one is slightly tailored for the egress IP use case.

alexanderConstantinescu commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1306,25 +1321,20 @@ func (oc *Controller) deleteEgressNode(egressNode *kapi.Node) error {
+				// workqueue's delta FIFO implementation will not trigger a
+				// watch event for updates on objects which have no semantic
+				// difference, hence: call the reconciliation function directly.
+				if err := oc.reconcileEgressIP(nil, &egressIP); err != nil {
```
Yes, because in `validateEgressIPStatus` we validate if the status on the `new` object is valid, and before executing `deleteEgressNode` we mark the node as not usable for egress IP assignment, meaning `validateEgressIPStatus` will see that one current status assignment is invalid and we will try to re-assign that status elsewhere.

alexanderConstantinescu commented on go-controller/pkg/ovn/egressip_test.go:
```diff
@@ -1591,7 +1591,360 @@ var _ = ginkgo.Describe("OVN master EgressIP Operations", func() {
-	ginkgo.Context("IPv6 on EgressIP UPDATE", func() {
+	ginkgo.Context("on EgressIP UPDATE", func() {
+
+		ginkgo.It("should update OVN on EgressIP spec change", func() {
```
> include what the spec change is?

OK

> Updating the Egress IPs with new values?

Yes

alexanderConstantinescu commented:
/retest

tssurya commented:
The noSnatGW job is failing for `Should validate the egress IP functionality against remote hosts` with:
```
Jan 19 00:32:42.724: FAIL: Step  15. Check connectivity from the remaining pod to an external "node" and verify that the IP is the node IP, failed, err: timed out waiting for the condition
Unexpected error:
    <*errors.errorString | 0xc00033f160>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
occurred
```
I am trying to review this commit https://github.com/ovn-org/ovn-kubernetes/pull/2734/commits/f2d7edacaf07cce4a4150d51d557fb3d80049180 specifically, @alexanderConstantinescu could you please add more details to the commit message in what's the bug with the fix we did in https://github.com/ovn-org/ovn-kubernetes/pull/2725 and why we need this commit that you are doing?

alexanderConstantinescu commented:
> The noSnatGW job is failing for `Should validate the egress IP functionality against remote hosts` with:

Fixed now.

> I am trying to review this commit [f2d7eda](https://github.com/ovn-org/ovn-kubernetes/commit/f2d7edacaf07cce4a4150d51d557fb3d80049180) specifically, @alexanderConstantinescu could you please add more details to the commit message in what's the bug with the fix we did in #2725 and why we need this commit that you are doing?

I added a more developed commit message
